File: go/src/algorithms/binary_search/binary_search.go
package binary_search

func search(nums []int, target int) int {
	lo := 0
	hi := len(nums)
	for lo < hi {
		mid := lo + (hi-lo)/2
		if nums[mid] > target {
			hi = mid
		} else if nums[mid] < target {
			lo = mid + 1
		} else {
			return mid
		}
	}
	return -1
}

File: ocaml/lib/algorithms/binary_search.ml
(** [binary_search v arr] returns the index of [v] in [arr] if it exists,
    and raises [Not_found] otherwise. [arr] must be sorted in ascending order
    and contain no duplicates. *)
let binary_search v arr =
  let rec search v lo hi =
    let mid = lo + (hi - lo) / 2 in
    if lo > hi then
      raise Not_found
    else if v > arr.(mid) then
      search v (mid + 1) hi
    else if v < arr.(mid) then
      search v lo mid
    else mid
  in
  let lo = 0 in 
  let hi = Array.length arr - 1 in
  search v lo hi

